### PR TITLE
CLOUDP-46347: Allow administrators to limit sizes

### PR DESCRIPTION
**Pull request by mrgunior**
Administrators are now able to adjust what plans/providers are available in the marketplace. 
This means an administrator can create their own providers.json file. I've also added ([providersConfigExample.json](https://github.com/mongodb/mongodb-atlas-service-broker/blob/CLOUDP-46347-allow-administrators-to-limit-sizes/providersConfigExample.json)) file as an example on how one can create their own and pass it as an `environment variable` outside of kubernetes. This will give administrators more control on what plans/sizes users can provision.

How it works:
- If you don't pass the environment variable `PROVIDERS_CONFIG_FILE`, no file will be used and the service broker will just return everything from atlas (providers with their plans). If you do choose to pass the environment variable, that will be used. The functionality is quite flexible. You are able to for example limit GCP sizes but use all from AWS, or vice verse. This also holds for the others (TENANT, AZURE).
- An administrator can adjust the file without having to restart anything. 
- I also added some asserts to test this this new addition and also readjusted the way we used to provision M2 and M5 sizes. Now everything is in 1 single file called [meta.go](https://github.com/mongodb/mongodb-atlas-service-broker/blob/CLOUDP-46347-allow-administrators-to-limit-sizes/pkg/atlas/meta.go) and causes less overhead.

### CODE REVIEW

File: pkg/atlas/meta.go [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
 	"fmt"
+	"io/ioutil"
 	"net/http"
 )
 
+// PathToProviderJSON is the path to the json file
+var PathToProviderJSON = "providers.json"
+
+// ProvidersFile represents all of the providers that have been set in the provider.json file
+type ProvidersFile struct {
```
Comment (antonlisovenko) [inline pkg/atlas/meta.go]: [nit] may be `ProvidersConfig` or whatever? The `File` suffix doesn't seem very clear to me

Comment (BenElgar) [reply pkg/atlas/meta.go]: Agreed.

File: pkg/atlas/meta.go [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
 
 import (
+	"encoding/json"
 	"fmt"
+	"io/ioutil"
 	"net/http"
 )
 
+// PathToProviderJSON is the path to the json file
+var PathToProviderJSON = "providers.json"
```
Comment (BenElgar) [inline pkg/atlas/meta.go]: This should be passed in by the user as an environment variable, marshalled into a meaningful object and then passed into the initialisation of the Atlas client.

File: pkg/atlas/providers.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (BenElgar) [inline pkg/atlas/providers.json]: File shouldn't be required.

File: pkg/atlas/meta.go [hunk 4]
```diff
@@ -21,8 +31,44 @@ type InstanceSize struct {
...
+	if isSet {
+		return &provider, err
+	}
+
+	// Otherwise we make a request to the private atlas API when it's not set
+	err = c.requestPrivate(http.MethodGet, path, nil, &provider)
 	return &provider, err
 }
+
+// setupProviders will setup all of the available providors together with their plans/sizes.
```
Comment (BenElgar) [inline pkg/atlas/meta.go]:
```suggestion
// setupProviders will setup all of the available providers together with their plans/sizes.
```

File: pkg/atlas/meta.go [hunk 5]
```diff
@@ -21,8 +31,44 @@ type InstanceSize struct {
...
+// Users may want to set their local providor, and combine that with a remote provider that isn't set locally.
+func setupProviders(name string) (provider Provider, isSet bool, err error) {
```
Comment (BenElgar) [inline pkg/atlas/meta.go]: This is a bad name on the grounds that this function does not set up providers; "set up" implies some side effect when this function doesn't have any and "providers" is plural when this function takes a single provider name.

File: pkg/atlas/meta.go [hunk 6]
```diff
@@ -21,8 +31,44 @@ type InstanceSize struct {
...
+		return &provider, err
+	}
+
+	// Otherwise we make a request to the private atlas API when it's not set
+	err = c.requestPrivate(http.MethodGet, path, nil, &provider)
 	return &provider, err
 }
+
+// setupProviders will setup all of the available providors together with their plans/sizes.
+// Users may want to set their local providor, and combine that with a remote provider that isn't set locally.
```
Comment (BenElgar) [inline pkg/atlas/meta.go]:
```suggestion
// Users may want to set their local provider, and combine that with a remote provider that isn't set locally.
```

File: pkg/atlas/providers.json [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    "providers" : [
+        {
+            "UNSET": {
```
Comment (BenElgar) [inline pkg/atlas/providers.json]: I don't really understand the point of the "SET" or "UNSET" objects. If the provider is included within the file then it should be used. If it shouldn't be used then it shouldn't be included within the file.

File: providers.json [hunk 8]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (BenElgar) [inline providers.json]: Why is this duplicated?

File: pkg/atlas/providers.json [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+                "@provider": "AZURE",
+                "InstanceSizes": {
+                    "M10": {"name": "M10"},
+                    "M20": {"name": "M20"},
+                    "M200": {"name": "M200"}
+                }
+            }
+        }
+    ]
+  }
```
Comment (BenElgar) [inline pkg/atlas/providers.json]: New line pls.

File: pkg/atlas/meta.go [hunk 10]
```diff
@@ -1,10 +1,20 @@
...
+	"io/ioutil"
 	"net/http"
 )
 
+// PathToProviderJSON is the path to the json file
+var PathToProviderJSON = "providers.json"
+
+// ProvidersFile represents all of the providers that have been set in the provider.json file
+type ProvidersFile struct {
+	Providers []map[string]Provider `json:"providers,omitempty"`
```
Comment (BenElgar) [inline pkg/atlas/meta.go]: I would argue that we shouldn't make the user conform to the structure of the undocumented APIs; the structure is non-obvious, unfamiliar and unnecessarily verbose. We also shouldn't override the provider object as received from the API—whilst at the moment we only use the instance sizes, there is a bunch of other information that we will hopefully use in the near future, such as costs and other plan metadata.

**Comment by BenElgar** [comment]
We also need a suite of tests to ensure this behaves as expected.

**Comment by BenElgar** [comment]
@antonlisovenko I agree we shouldn't have to expect the user to enter the container and modify this file. Instead we should allow users to specify the path to this file as an environment variable and then they can mount the file as a secret. An extension could even be to watch the file so that users don't have to restart the pod when changing the available plans (though definitely not in this PR).

The environment variables would work but there are 4 providers, each with about 25 different plans, so we'd end up with about 100 environment variables that users would have to individually specify and we'd have to document.

**Comment by antonlisovenko** [comment]
I forgot that the user controls the deployment by himself.
In this case I think we even can skip the env variable for the path and just use the predefined path where the json is expected to be mounted. The user will specify it as a mount point then. 
What do you think @BenElgar ?

**Comment by BenElgar** [comment]
I'd still like to provide an environment variable to select the file as it shouldn't be required: if no file is specified then it should allow any size Atlas cluster.

File: test/integration/integration_test.go [hunk 11]
```diff
@@ -140,6 +145,40 @@ func TestProvision(t *testing.T) {
...
+		RawParameters: []byte(params),
+	}, true)
+	assert.Error(t, atlas.ErrPlanIDNotFound, err)
+
+	_, err = broker.Provision(ctx, instanceID, brokerapi.ProvisionDetails{
+		ServiceID:     "aosb-cluster-service-azure",
+		PlanID:        "aosb-cluster-plan-azure-m60",
+		RawParameters: []byte(params),
+	}, true)
+	assert.Error(t, atlas.ErrPlanIDNotFound, err)
```
Comment (BenElgar) [inline test/integration/integration_test.go]: Can these be part of a separate integration test to keep the responsibilities clear. We also need a test for what happens if no file is provided.

File: main.go [hunk 12]
```diff
@@ -75,14 +75,17 @@ Docker Image: quay.io/mongodb/mongodb-atlas-service-broker`
 }
 
 func startBrokerServer() {
+	// Administrators can control what providers/plans are available to users
+	providersConfig := getEnvOrDefault("PROVIDERS_CONFIG_FILE", "")
```
Comment (BenElgar) [inline main.go]: If the user provides an environment variable that doesn't point at a valid JSON file we should panic.

File: pkg/atlas/meta.go (hunk 2 above)
Comment (antonlisovenko) [reply pkg/atlas/meta.go]: do we need this field in the client? Could we just call `providersConfig := getEnvOrDefault("PROVIDERS_CONFIG_FILE", "")` from `getProvider` and avoiding quite unnecessary state into the client?

File: pkg/atlas/meta.go [hunk 13]
```diff
@@ -20,9 +25,55 @@ type InstanceSize struct {
...
+
+	// Otherwise we make a request to the private atlas API when it's not set
+	path := fmt.Sprintf("cloudProviders/%s/options", name)
+	err = c.requestPrivate(http.MethodGet, path, nil, &provider)
 	return &provider, err
 }
+
+// getLocalProvider will setup all of the available providors together with their plans/sizes.
+// Users may want to set their local provider, and combine that with a remote provider that isn't set locally.
+func getLocalProvider(name string, pathToFile string) (provider Provider, isSet bool, err error) {
```
Comment (antonlisovenko) [inline pkg/atlas/meta.go]: We need to check against the situation when the env variable is not set (`pathToFile` is empty) and return `isSet=false` in this case

File: test/integration/integration_test.go [hunk 14]
```diff
@@ -282,9 +321,18 @@ func TestUpdate(t *testing.T) {
...
-		PlanID:        "aosb-cluster-plan-aws-m20",
+		PlanID:        "aosb-cluster-plan-aws-m60",
+		RawParameters: []byte(params),
+	}, true)
+
+	assert.Error(t, atlas.ErrPlanIDNotFound, err)
+
+	_, err = broker.Update(ctx, instanceID, brokerapi.UpdateDetails{
+		ServiceID:     "aosb-cluster-service-aws",
+		PlanID:        "aosb-cluster-plan-aws-m10",
```
[the remark this answers is not in the document]
Comment (mrgunior) [reply test/integration/integration_test.go]: You're right, I changed the assert instead of the Update()

File: providersConfigExample.json [hunk 15]
```diff
@@ -0,0 +1,6 @@
+{
+    "GCP": {"InstanceSizes": ["M10", "M20", "M200"]},
+    "AWS": {"InstanceSizes": ["M10", "M10", "M100", "M140", "M140"]},
+    "AZURE": {"InstanceSizes": ["M10","M20","M200"]},
+    "TENANT": {"InstanceSizes": ["M2","M5"]}
+}
```
Comment (BenElgar) [inline providersConfigExample.json]: Missing new line. Also this file should live in `/samples`.

**Comment by mrgunior** [comment]
I would suggest not to merge this yet, there is a bug in the implementation that wasn't  there in the previous commit. My suggestion would be to either revamp how catalog.go and meta.go interact with eachother, or have it how the previous commit was. Except that we combine it with how the json file is read, and passed to the broker in this current commit.

**Pull request merged**